source/extensions/filters/http/grpc_field_extraction/filter.cc

```diff
@@ -0,0 +1,229 @@
+                     proto_path.status().ToString());
+    auto& status = proto_path.status();
+    rejectRequest(status.raw_code(), status.message(),
+                  generateRcDetails(kRcDetailFilterGrpcFieldExtraction,
+                                    absl::StatusCodeToString(status.code()),
+                                    kRcDetailErrorTypeBadRequest));
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+
+  auto* extractor = filter_config_.FindExtractor(*proto_path);
+  if (!extractor) {
+    ENVOY_STREAM_LOG(debug, "gRPC method `{}` isn't configured for field extraction",
+                     *decoder_callbacks_, *proto_path);
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  extractor_ = extractor;
+  request_msg_converter_ = std::make_unique<MessageConverter>(
+      std::make_unique<std::function<std::unique_ptr<Protobuf::field_extraction::MessageData>()>>(
+          []() { return std::make_unique<Protobuf::field_extraction::CordMessageData>(); }),
+      decoder_callbacks_->decoderBufferLimit());
+
+  return Envoy::Http::FilterHeadersStatus::StopIteration;
+}
+
+Envoy::Http::FilterDataStatus Filter::decodeData(Envoy::Buffer::Instance& data, bool end_stream) {
+  ENVOY_STREAM_LOG(debug, "decodeData: data size={} end_stream={}", *decoder_callbacks_,
+                   data.length(), end_stream);
+  if (!extractor_ || extraction_done_) {
+    return Envoy::Http::FilterDataStatus::Continue;
+  }
+
+  if (auto status = handleDecodeData(data, end_stream); !status.got_messages) {
+    return status.filter_status;
+  }
+
+  return Envoy::Http::FilterDataStatus::Continue;
+}
+
+Filter::HandleDecodeDataStatus Filter::handleDecodeData(Envoy::Buffer::Instance& data,
+                                                        bool end_stream) {
+  ABSL_DCHECK(extractor_ && request_msg_converter_);
+
+  auto buffering = request_msg_converter_->accumulateMessages(data, end_stream);
+  if (!buffering.ok()) {
+    const absl::Status& status = buffering.status();
+    rejectRequest(status.raw_code(), status.message(),
+                  generateRcDetails(kRcDetailFilterGrpcFieldExtraction,
+                                    absl::StatusCodeToString(status.code()),
+                                    kRcDetailErrorRequestBufferConversion));
+    return HandleDecodeDataStatus(Envoy::Http::FilterDataStatus::StopIterationNoBuffer);
+  }
+
+  if (buffering->empty()) {
+    ENVOY_STREAM_LOG(debug, "not a complete msg", *decoder_callbacks_);
+    // Not a complete message.
+    return HandleDecodeDataStatus(Envoy::Http::FilterDataStatus::StopIterationNoBuffer);
+  }
+
+  // Buffering returns a list of messages.
+  for (size_t msg_idx = 0; msg_idx < buffering->size(); ++msg_idx) {
+    std::unique_ptr<StreamMessage> message_data = std::move(buffering->at(msg_idx));
+
+    // MessageConverter uses a empty StreamMessage to denote the end.
+    if (message_data->size() == -1) {
```

Comment:
Why -1 and not 0?

Capture sentinel values as constants at top of this file. Or move the constant definition into PFX library and import that directly. Currently this is a magic number

Reply:
0 means empty, -1 means null. Anyway, change to use message() which is more intuitive.